go/tools/gazelle/merger/merger.go

```diff
@@ -0,0 +1,232 @@
+			continue
+		}
+		if literal(c.X) == "load" {
+			mergeLoad(c, other, f)
+		} else {
+			merge(c, other)
+		}
+	}
+	f.Stmt = append(f.Stmt, newStmt...)
+	return f, nil
+}
+
+// merge takes new info from src and merges into dest.
+// pre: these calls are the same X and 'name'
+func merge(src, dest *bzl.CallExpr) {
+	for _, item := range src.List {
+		nkv, ok := item.(*bzl.BinaryExpr)
+		if !ok {
+			continue
+		}
+		k := literal(nkv.X)
+		if !mergeableFields[k] {
+			continue
+		}
+		// TODO(pmbethe09): allow '# keep' on src files.
+		replace(dest, nkv)
+	}
+}
+
+func replace(dest *bzl.CallExpr, update *bzl.BinaryExpr) {
+	k := literal(update.X)
+	for i, item := range dest.List {
+		okv, ok := item.(*bzl.BinaryExpr)
+		if !ok {
+			continue
+		}
+		if k == literal(okv.X) {
+			dest.List[i] = update
+			return
+		}
+	}
+	dest.List = append(dest.List, update)
+}
+
+func mergeLoad(src, dest *bzl.CallExpr, oldfile *bzl.File) {
+	vals := make(map[string]bzl.Expr)
+	for _, v := range src.List[1:] {
+		vals[stringValue(v)] = v
+	}
+	for _, v := range dest.List[1:] {
+		rule := stringValue(v)
+		if _, ok := vals[rule]; !ok && ruleUsed(rule, oldfile) {
+			vals[rule] = v
+		}
+	}
+	keys := make([]string, 0, len(vals))
+	for k := range vals {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	dest.List = dest.List[:1]
+	for _, k := range keys {
+		dest.List = append(dest.List, vals[k])
+	}
+}
+
+func ruleUsed(rule string, oldfile *bzl.File) bool {
+	for _, s := range oldfile.Stmt {
+		c, ok := s.(*bzl.CallExpr)
+		if !ok {
+			continue
+		}
+		if rule == literal(c) {
+			return true
+		}
+	}
+	return false
+}
+
+// match looks for the matching CallExpr in f using X and name
+// i.e. two 'go_library(name = "foo", ...)' are considered matches
+// despite the values of the other fields.
+// exception: if c is a 'load' statement, the match is done on the first value.
+func match(f *bzl.File, c *bzl.CallExpr) (*bzl.CallExpr, error) {
+	x := literal(c.X)
+	if x == "" {
+		return nil, fmt.Errorf("expected LiteralExpr, got; %v", c)
+	}
+	var m matcher
+	if x == "load" {
+		if len(c.List) == 0 {
+			return nil, nil
+		}
+		m = &loadMatcher{stringValue(c.List[0])}
+	} else {
+		m = &nameMatcher{x, get("name", c)}
+	}
+	for _, s := range f.Stmt {
+		other, ok := s.(*bzl.CallExpr)
+		if !ok {
+			continue
+		}
+		if m.match(other) {
+			return other, nil
+		}
+	}
+	return nil, nil
+}
+
+type matcher interface {
+	match(c *bzl.CallExpr) bool
+}
+
+type nameMatcher struct {
+	x, name string
+}
+
+func (m *nameMatcher) match(c *bzl.CallExpr) bool {
+	return m.x == literal(c.X) && m.name == get("name", c)
+}
+
+type loadMatcher struct {
+	load string
+}
+
+func (m *loadMatcher) match(c *bzl.CallExpr) bool {
+	return literal(c.X) == "load" && len(c.List) > 0 && m.load == stringValue(c.List[0])
+}
+
+// get returns the value for the 'name = value' entry in the calls arguments
+// or "" if not found.
+func get(name string, c *bzl.CallExpr) string {
```

Comment:
These accessor functions are basically reinventing the accessors in `bzl.Rule`.
Why don't you wrap your `bzl.CallExpr` with `bzl.Rule` and reuse the accessors in `bzl.Rule`?